pkg/util/metrics.go

```diff
@@ -0,0 +1,182 @@
+	return strings.Replace(prefix+name, "-", "_", -1)
+}
+
+// Best effort metric registration with Prometheus.
+func RegisterMetric(metric prometheus.Collector) {
+	if err := prometheus.Register(metric); err != nil {
+		glog.Warningf("Error while registering Prometheus metric: [%v]", err)
+	}
+}
+
+type PositiveGauge struct {
+	mux         sync.RWMutex
+	name        string
+	gaugeMetric *prometheus.GaugeVec
+}
+
+// Gauge with conditional decrement ensuring its value is never negative.
+func NewPositiveGauge(name string, description string, labels []string) *PositiveGauge {
+	for i, label := range labels {
+		labels[i] = CreateValidMetricNameLabel("", label)
+	}
+
+	gauge := prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: name,
+			Help: description,
+		},
+		labels,
+	)
+
+	RegisterMetric(gauge)
+
+	return &PositiveGauge{
+		gaugeMetric: gauge,
+		name:        name,
+	}
+}
+
+func fetchGaugeValue(m *prometheus.GaugeVec, labels map[string]string) float64 {
+	// Hack to get the current value of the metric to support PositiveGauge
+	pb := &prometheus_model.Metric{}
+
+	m.With(labels).Write(pb)
+	return pb.GetGauge().GetValue()
+}
+
+func (c *PositiveGauge) Value(labelMap map[string]string) float64 {
+	c.mux.RLock()
+	defer c.mux.RUnlock()
+	return fetchGaugeValue(c.gaugeMetric, labelMap)
+}
+
+// Increment the Metric for the labels specified
+func (c *PositiveGauge) Inc(labelMap map[string]string) {
+
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	if m, err := c.gaugeMetric.GetMetricWith(labelMap); err != nil {
+		glog.Errorf("Error while posting metrics: %v", err)
+
+	} else {
+		glog.V(2).Infof("Incrementing %s with labels %s", c.name, labelMap)
+		m.Inc()
+	}
+}
+
+// Decrement the metric only if its positive for the labels specified
+func (c *PositiveGauge) Dec(labelMap map[string]string) {
+
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	// Decrement only if positive
+	val := fetchGaugeValue(c.gaugeMetric, labelMap)
+	if val > 0 {
+		glog.V(2).Infof("Decrementing %s with labels %s metricVal to %v", c.name, labelMap, val-1)
+		if m, err := c.gaugeMetric.GetMetricWith(labelMap); err != nil {
+			glog.Errorf("Error while posting metrics: %v", err)
+		} else {
+			m.Dec()
+		}
+	}
+}
+
+type WorkQueueMetrics struct {
+	prefix string
+}
+
+func InitializeMetrics(endpoint string, port string, prefix string) {
+
+	// Start the metrics endpoint for Prometheus to scrape
+	http.Handle(endpoint, promhttp.Handler())
+	go http.ListenAndServe(port, nil)
+	glog.Infof("Started Metrics server at localhost%s%s", port, endpoint)
+
+	workQueueMetrics := WorkQueueMetrics{prefix: prefix}
+	workqueue.SetProvider(&workQueueMetrics)
+}
+
+// Depth Metric for kubernetes workqueue
+func (p *WorkQueueMetrics) NewDepthMetric(name string) workqueue.GaugeMetric {
+	depthMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: CreateValidMetricNameLabel(p.prefix, name+"_depth"),
+		Help: "Current depth of workqueue: " + name,
```

Comment:
Can we use `fmt.Sprintf` for formatting the `Help` strings?

Reply:
done